### DISCUSSION TITLE
Does Mono on Mac or Linux have the same compiling restriction as .NET Core?

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#107] [Discussions]
Is compiling available on Mono? That is, does Mono also restrict use of Reflection.Emit in terms of saving assemblies to the harddrive? If not then why even bother with .NET Core?

I ask because until today I didn't even really try to run my .NET 4.8 app on Mono, as I was focused on .NET Core 5.0. I saw that on my new Mac (running windows on Parallels) IronScheme.Console-v4.exe runs fine via Mono, so on a whim I tried my .NET console app (embedding IronScheme) targeting .NET 4.8 on Mono, and it ran. So, being lazy, I thought I'd ask before I test it myself. 

My follow up question would be: what limitations in general are there on Mono running .NET 4.8 binaries?

I'm enjoying .NET a lot with my combination of F# and IronScheme but I have not been clear, since returning from a several year hiatus from .NET, about all the CLR runtime targets once one is no longer on a Windows machine. I was for some reason astonished that my Windows console app would run as-is on Mono. So now I want to push that envelope but would appreciate any hints where the dragons lie, from those who have done more of this "portable .NET thing" than I have up to this point.

**Comment by bcalco** [comment]
@leppie I'd love a concrete example of compiling a script/library so I can try to reproduce compiling and running it in my app's wrapping of IronScheme, both on Windows and now on Mac via Mono. If the latter doesn't work it's OK, I still would like it for Windows.

**Comment by leppie** [comment]
Last time I checked, it did work, but there may be issues. I dont really keep track of it.

The simplest way to check if compiling work is just to run `(compile-system-libraries)` from the REPL (ideally from a brand new release zip). 

It takes about a minute on my PC using .NET 4.8. 64-bit. 

The exit the REPL, load it again, and run `(compile-system-libraries)`. This still should just load all the precompiled libraries without compiling anyone again. If it recompiles anything, something is not right.

This step takes about 12 seconds on my PC using .NET 4.8. 64-bit. 

Also, compiling for CLR v2 32-bit is generally the safest if you want to reuse them in any upstream versions.

You can see if you did what I did above, the libraries wont work on CLR v2 32-bit.

**Reply by bcalco** [reply]
that's helpful, @leppie, thanks; I'll give it a spin on mono today as you describe. What I'm also looking for is an example of compiling a custom library or script, where to put stuff, what the incantation should be, where to expect what output, that kind of detail. Because if I can, I will generally want to, particularly for libraries I want to add to my application (which is embedding IronScheme). So far I'm not having any problem runnin my app's tests on Mono, and the tests are not trivial - including all the "Little" books and most of TSPL4, as well as some miscellaneous tests involving SRFI 219. They all run beautifully, and quite fast, actually. So I'm not obsessed with optimizing via compiling or anything like that, but for some libraries I'd prefer an internal/binary implementation if available.

**Reply by bcalco** [reply]
OK so here was what happened on my Mac:

```
adsmactf01@ADSMACTF01s-MBP IronScheme % mono IronScheme.Console-v4.exe
IronScheme 1.0.243-2f84779 github.com/IronScheme © 2007-2019 Llewellyn Pritchard (Mono .NET 4.6 64-bit)
> (compile-system-libraries)
compiling ironscheme.clr.helpers.dll
compiling ironscheme.clr.dll
* Assertion at sre-save.c:1791, condition `mono_class_is_ginst (m->klass) || mono_class_is_gtd (m->klass)' not met


=================================================================
	Native Crash Reporting
=================================================================
Got a abrt while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
	Native stacktrace:
=================================================================
	0x1086de719 - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : mono_dump_native_crash_info
	0x1086765ee - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : mono_handle_native_crash
	0x1086ddcdf - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : sigabrt_signal_handler
	0x7ff8169a7e2d - /usr/lib/system/libsystem_platform.dylib : _sigtramp
	0x1 - Unknown
	0x7ff8168ded10 - /usr/lib/system/libsystem_c.dylib : abort
	0x1088caf57 - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : monoeg_assert_abort
	0x1088abdbf - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : mono_log_write_logfile
	0x1088cb3ee - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : monoeg_g_logv_nofree
	0x1088cb56f - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : monoeg_assertion_message
	0x1088cb5aa - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : mono_assertion_message
	0x108832df2 - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : fixup_method
	0x1087dcda0 - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : mono_g_hash_table_foreach
	0x108831c25 - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : mono_image_build_metadata
	0x108796329 - /Library/Frameworks/Mono.framework/Versions/6.12.0/bin/mono-sgen64 : ves_icall_ModuleBuilder_build_metadata_raw
	0x10f1a14ca - Unknown
	0x10c882a86 - /Library/Frameworks/Mono.framework/Versions/6.12.0/lib/mono/4.5/mscorlib.dll.dylib : System_Reflection_Emit_AssemblyBuilder_Save_string_System_Reflection_PortableExecutableKinds_System_Reflection_ImageFileMachine
	0x10e0ff4f3 - Unknown
	0x10ad99e3b - Unknown
	0x10ad999b9 - Unknown
	0x108f86a3b - Unknown
	0x10f19e1a3 - Unknown

=================================================================
	Telemetry Dumper:
=================================================================
Pkilling 0x123145454215168x from 0x4453705216x
Entering thread summarizer pause from 0x4453705216x
Finished thread summarizer pause from 0x4453705216x.
zsh: abort      mono IronScheme.Console-v4.exe
adsmactf01@ADSMACTF01s-MBP IronScheme % Mono process hang detected, sending kill signal to pid 60324
```

**Reply by bcalco** [reply]
It did successfully generate and save the ironscheme.helpers.dll, but something about the ironscheme.clr code caused it to kaboom. I will try the v2 next. If perchance it works, and I re-enter v4 64 bit REPL, will they get loaded? How will I know if the REPL is automatically loading them?

**Reply by bcalco** [reply]
OK, v2 32 bit gave the same as v4 64 and 32 bit, except for this nuance:

```
adsmactf01@ADSMACTF01s-MBP IronScheme % mono IronScheme.Console32-v2.exe
WARNING: The runtime version supported by this application is unavailable.
Using default runtime: v4.0.30319
IronScheme 1.0.243-2f84779 github.com/IronScheme © 2007-2019 Llewellyn Pritchard (Mono .NET 4.6 64-bit)
> (compile-system-libraries)
compiling ironscheme.clr.helpers.dll
compiling ironscheme.clr.dll
```

I can try to install the v2 framework in my Mono installation if you think it will help but I'm dubious at this point. Same exact error when it kabooms.

**Reply by leppie** [reply]
That's Mono for you :D 

You can try log an issue with them, but from my experience, bugs took multiple years to be fixed, if at all. I gave up a long time ago. Also Miguel does not like me very much.

That said, the fact that it actually loaded all the libraries, means that that it should run just fine.

Something else you can try is compile them all on Windows with .NET desktop, then copy them to the dir on Mac/Linux. 

Then run it again. If it loads everything without crashing or trying to recompile a library, it should be fine.

This is where .NET Core fails for some reason, it wants to recompile, and I dont know why. Perhaps it will be fine with Mono.

**Reply by leppie** [reply]
> How will I know if the REPL is automatically loading them?

I think the `--show-loaded-libraries` command line might work. 

If it cant load from a precompiled library, it will show a warning why, and then recompile it. You should clean up stale libraries as I recall it could cause problems.

**Comment by bcalco** [comment]
I will say, if this can be made to work on Mono, I'm happy to suggest my users use Mono for the extra boost. But I will turn my focus to living with the limitations of .NET Core in this regard. It absolutely must run on Linux and Mac (in that order of importance).

**Reply by bcalco** [reply]
And again for the record I've seen no problem running on Mono generally.

**Reply by leppie** [reply]
.NET Core should generally be a lot faster than Mono. Even vs .NET desktop. 

You can run the R6RS test suite by running `(ironscheme-test)` (note there are 3/4 expected failing tests, but this is a quick benchmark).
